org.eclipse.jdt.core/model/org/eclipse/jdt/core/JavaCore.java

```diff
@@ -5971,6 +5971,75 @@ public static void run(IWorkspaceRunnable action, ISchedulingRule rule, IProgres
 			workspace.run(new BatchOperation(action), rule, IWorkspace.AVOID_UPDATE, monitor);
 		}
 	}
+	/**
+	 * @since 3.37
+	 */
+	@FunctionalInterface
+	public static interface JavaCallable<V, E extends Exception> {
+		/**
+		 * Computes a value or throws an exception.
+		 *
+		 * @return the result
+		 * @throws E the Exception of given type
+		 */
+		V call() throws E;
+	}
+	/**
+	 * @since 3.37
+	 */
+	@FunctionalInterface
+	public static interface JavaRunnable<E extends Exception> {
+		/**
+		 * Runs or throws an exception.
+		 *
+		 * @throws E the Exception of given type
+		 */
+		void run() throws E;
+	}
+
+	/**
+	 * Calls the argument and returns its result or its Exception. To be used on read-only batch operations to cache
+	 * data.
+	 *
+	 * @param callable
+	 *            A JavaCallable that can throw an Exception
+	 * @return the result
+	 * @exception E
+	 *                An exception that is thrown by the callable.
+	 * @since 3.37
+	 */
+	public static <T, E extends Exception> T call(JavaCallable<T, E> callable) throws E {
+		JavaModelManager manager = JavaModelManager.getJavaModelManager();
+		Object instance = new Object();
+		boolean hadTemporaryCache = manager.hasTemporaryCache();
+		try {
+			manager.getTemporaryCache();
+			manager.cacheZipFiles(instance);
+			return callable.call();
+		} finally {
+			manager.flushZipFiles(instance);
+			if (!hadTemporaryCache) {
+				manager.resetTemporaryCache();
+			}
+		}
+	}
```

Comment:
Looking at this, it looks like a `JavaCore` knows too much about the internal workings of `JavaModelManager` (it needs to call 3 methods in the right order in order to make it work properly). I think it would be better to try and hide such details in order to avoid even more unnecessary coupling between these two (already pretty coupled) classes.
A "quick and dirty" solution would be to move everything around `callable.call()` to a new method inside the `JavaModelManager`, something like:

```suggestion
	public static <T, E extends Exception> T call(JavaCallable<T, E> callable) throws E {
		JavaModelManager manager = JavaModelManager.getJavaModelManager();
		Object instance = new Object();
		boolean hadTemporaryCache = manager.hasTemporaryCache();
		try {
			manager.getTemporaryCache();
			manager.cacheZipFiles(instance);
			return callable.call();
		} finally {
			manager.flushZipFiles(instance);
			if (!hadTemporaryCache) {
				manager.resetTemporaryCache();
			}
		}
	}
```

... and you could move `JavaCallable` & Co. down to the `JavaModelManager` too instead of exposing them as public API (they would be `internal`).

Reply:
> AFAICT every usage of this new method can get away with a Callable since there are no checked exceptions flying around, am I right?

I just saw that you added a few more calls to `run` / `call` in https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/933 so my last comment is probably already outdated (wrong). You may have to expose `JavaCallable` / `JavaRunnable` but the good thing is that is currently only needed in `eclipse.jdt.ui` which means you could still expose them as `internal` and add `e,jdt.ui` as a "friend of" `e.jdt.core`.

Reply:
I want public api so that others can use it too. for example PDE may useit.  I also saw  minor usecases where jdt-debug would also benefit.
JavaCore  was selected instead of JavaModelManager  as requested in https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1614#issuecomment-1822597670

Reply:
Not sure why my comment is unclear.

I meant: To be consistent, new public "run something on Java model" API methods should be added on JavaCore, next to existing API. It makes sense to keep similar API in same class.

Reply:
That's what I thought 👍 , which is good because then @jukzi can use my suggestion and keep the current API consistent without adding more coupling between `JavaCore` and `JavaModelManager`:
* The entry point is still a `run`-method in `JavaCore`
* `JavaCore` delegates to `JavaModelManager`, which knows its own internals and knows how to properly handle caching

Reply:
JavaCore now delegates to JavaModelManager